_posts/2023-04-30-writing-code-with-chatgpt-improve-it-with-kani.md

```diff
@@ -0,0 +1,369 @@
+The last task we explore in this blog is: combining Kani and ChatGPT to generate code that passes verification.
+
+Recall that we found the overflow error in the expression `a + b` in the generated function `integer_average` with Kani.
+We shared this information with ChatGPT.
+
+```
+Q: The expression (a + b) / 2 may overflow.
+```
+
+ChatGPT agreed with us and fixed the overflow with a new implementation using `checked_add`:
+
+``` rust
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    match num1.checked_add(num2) {
+        Some(sum) => sum / 2,
+        None => {
+            // Handle overflow gracefully
+            println!("Overflow occurred!");
+            std::i32::MAX // Fallback value
+        }
+    }
+}
+```
+
+We reran Kani on this new implementation.
+Kani verified that this new version is correct!
+However, the function is no longer complete because it doesn't computes the average for all `i32` integers as we specified.
+For example, it does not compute the average for the input `integer_average(std::i32::MAX, 1)`.
+So, we tried again with the following prompt to generate a complete version of `integer_average`.
+
+```
+Q: Can you give me a fix that does not throw any error?
+```
+
+This time, ChatGPT found us an implementation that accepts all `i32` integers as input.
+
+```
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    num1 / 2 + num2 / 2 + (num1 % 2 + num2 % 2) / 2
+}
+```
+
+Kani verifies that this implementation is correct.
+
+```
+> kani integer_average.rs
+
+...
+SUMMARY:
+ ** 0 of 6 failed
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.5309424s
+```
+
+
+
+## A More Practical Example: Program Optimization
+In this section, we will repeat the three tasks on a more practical and challenging problem.
+
+First, we wrote a Rust function that computes the lowest unset bit in a `u32` integer.
+```
+fn lowest_unset_bit_ori(x: u32) -> Option<u8> {
+    let mut mask = 1;
+    for i in 0..32 {
+        if x & mask == 0 {
+            return Some(i);
+        }
+        mask <<= 1;
+    }
+    None
+} 
+```
+Our implementation uses a loop to iterate through all 32 bits in the input.
+And we know that this could be slower than using some [bit twiddling hacks](https://graphics.stanford.edu/~seander/bithacks.html).
+In this case, our goal is to optimize the function with ChatGPT.
+This is the prompt we used, which includes the implementation we wrote:
+
+``` 
+Q: Optimize this Rust function:
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    let mut mask = 1;
+    for i in 0..32 {
+        if x & mask == 0 {
+            return Some(i);
+        }
+        mask <<= 1;
+    }
+    None
+} 
+```
+
+After a few seconds, ChatGPT gave us a solution
+
+``` rust
+fn lowest_unset_bit_opt(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    let y = x | (x + 1);
+    let z = y ^ (y + 1);
+    Some((z.trailing_zeros() % 32) as u8)
+}
+```
+
+along with an explanation.
+
+
```

Comment:
This section is now quite long - this looks like a great place to open a subsection with name "Equivalence Checking"